File: fe/lib/social/news/models/news_report.dart
class NewsReportReasons {
  const NewsReportReasons._();

  static const Map<String, String> labels = <String, String>{
    'spam': 'Spam o pubblicità',
    'harassment': 'Molestie o minacce',
    'hate': 'Odio o discriminazione',
    'privacy': 'Violazione della privacy',
    'illegal_content': 'Contenuto illecito',
    'other': 'Altro',
  };

  static bool isValid(String reason) => labels.containsKey(reason);

  static String labelOf(String reason) => labels[reason] ?? 'Altro';
}

class NewsReport {
  final int id;
  final String category;
  final String newsId;
  final int? groupId;
  final String conversationId;
  final int reporterUserId;
  final int? reportedUserId;
  final String reason;
  final String description;
  final String status;
  final String moderationAction;
  final String moderationNote;
  final int? reviewedByUserId;
  final DateTime? reviewedAt;
  final DateTime? disclosureConsentAt;
  final DateTime? disclosureOpenedAt;
  final DateTime createdAt;

  const NewsReport({
    required this.id,
    required this.category,
    required this.newsId,
    required this.groupId,
    required this.conversationId,
    required this.reporterUserId,
    required this.reportedUserId,
    required this.reason,
    required this.description,
    required this.status,
    required this.moderationAction,
    required this.moderationNote,
    required this.reviewedByUserId,
    required this.reviewedAt,
    required this.disclosureConsentAt,
    required this.disclosureOpenedAt,
    required this.createdAt,
  });

  bool get isPending => status == 'pending' || status == 'under_review';

  bool get hasDisclosureConsent => disclosureConsentAt != null;

  bool get wasDisclosureOpened => disclosureOpenedAt != null;

  String get reasonLabel => NewsReportReasons.labelOf(reason);

  factory NewsReport.fromJson(Map<String, dynamic> json) {
    return NewsReport(
      id: _toInt(json['id']) ?? 0,
      category: json['category']?.toString().trim() ?? '',
      newsId: json['news_id']?.toString().trim() ?? '',
      groupId: _toInt(json['group_id']),
      conversationId: json['conversation_id']?.toString().trim() ?? '',
      reporterUserId: _toInt(json['reporter_user_id']) ?? 0,
      reportedUserId: _toInt(json['reported_user_id']),
      reason: json['reason']?.toString().trim() ?? 'other',
      description: json['description']?.toString().trim() ?? '',
      status: json['status']?.toString().trim() ?? 'pending',
      moderationAction:
          json['moderation_action']?.toString().trim() ?? 'none',
      moderationNote: json['moderation_note']?.toString().trim() ?? '',
      reviewedByUserId: _toInt(json['reviewed_by_user_id']),
      reviewedAt: _parseNullableDate(json['reviewed_at']),
      disclosureConsentAt: _parseNullableDate(json['disclosure_consent_at']),
      disclosureOpenedAt: _parseNullableDate(json['disclosure_opened_at']),
      createdAt: _parseDate(json['created_at']),
    );
  }
}

class NewsReportListResult {
  final List<NewsReport> items;
  final int total;

  const NewsReportListResult({
    required this.items,
    required this.total,
  });

  factory NewsReportListResult.fromJson(Map<String, dynamic> json) {
    final dynamic rawItems = json['items'];

    final List<NewsReport> items = rawItems is List
        ? rawItems
            .whereType<Map>()
            .map(
              (Map<dynamic, dynamic> item) => NewsReport.fromJson(
                Map<String, dynamic>.from(item),
              ),
            )
            .toList()
        : <NewsReport>[];

    return NewsReportListResult(
      items: items,
      total: _toInt(json['total']) ?? items.length,
    );
  }
}

class NewsReportDisclosure {
  final int reportId;
  final String category;
  final String newsId;
  final int? authorId;
  final String authorName;
  final DateTime? createdAt;
  final String content;
  final bool verified;
  final List<String> wrapTargets;

  const NewsReportDisclosure({
    required this.reportId,
    required this.category,
    required this.newsId,
    required this.authorId,
    required this.authorName,
    required this.createdAt,
    required this.content,
    required this.verified,
    required this.wrapTargets,
  });

  factory NewsReportDisclosure.fromJson(Map<String, dynamic> json) {
    final dynamic targets = json['wrap_targets'];

    return NewsReportDisclosure(
      reportId: _toInt(json['report_id']) ?? 0,
      category: json['category']?.toString().trim() ?? '',
      newsId: json['news_id']?.toString().trim() ?? '',
      authorId: _toInt(json['author_id']),
      authorName: json['author_name']?.toString().trim() ?? '',
      createdAt: _parseNullableDate(json['created_at']),
      content: json['content']?.toString() ?? '',
      verified: json['verified'] == true,
      wrapTargets: targets is List
          ? targets.map((dynamic item) => item.toString()).toList()
          : <String>[],
    );
  }
}

int? _toInt(dynamic value) {
  if (value is int) {
    return value;
  }
  if (value is num) {
    return value.toInt();
  }
  return int.tryParse(value?.toString() ?? '');
}

DateTime _parseDate(dynamic value) {
  final DateTime? parsed = DateTime.tryParse(value?.toString() ?? '');
  if (parsed == null) {
    throw const FormatException('Data non valida.');
  }
  return parsed.toLocal();
}

DateTime? _parseNullableDate(dynamic value) {
  final String text = value?.toString().trim() ?? '';
  if (text.isEmpty) {
    return null;
  }
  return DateTime.tryParse(text)?.toLocal();
}

File: BE/schemas/news_report.py
from datetime import datetime

from typing import Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    field_validator,
    model_validator,
)


NewsReportCategory = Literal[
    "avvisi",
    "gruppi",
    "private",
]

NewsReportReason = Literal[
    "spam",
    "harassment",
    "hate",
    "privacy",
    "illegal_content",
    "other",
]

NewsReportStatus = Literal[
    "pending",
    "under_review",
    "resolved",
    "dismissed",
]

NewsReportAction = Literal[
    "none",
    "hide_news",
    "remove_news",
]


class NewsReportCreate(
    BaseModel,
):
    category: NewsReportCategory

    news_id: str = Field(
        min_length=1,
        max_length=64,
    )

    reason: NewsReportReason

    description: str = Field(
        default="",
        max_length=2000,
    )

    group_id: int | None = Field(
        default=None,
        gt=0,
    )

    other_user_id: int | None = Field(
        default=None,
        gt=0,
    )

    disclosed_content_key: str | None = Field(
        default=None,
        max_length=512,
    )

    disclosure_consent: bool = False

    @field_validator(
        "news_id",
    )
    @classmethod
    def validate_news_id(
        cls,
        value: str,
    ) -> str:
        normalized = value.strip()

        if not normalized or "/" in normalized:
            raise ValueError(
                "Identificativo della news non valido.",
            )

        return normalized

    @field_validator(
        "description",
    )
    @classmethod
    def normalize_description(
        cls,
        value: str,
    ) -> str:
        return value.strip()

    @model_validator(
        mode="after",
    )
    def validate_target(
        self,
    ):
        if self.category == "gruppi" and self.group_id is None:
            raise ValueError(
                "Per segnalare una news di gruppo serve il gruppo.",
            )

        if self.category == "private" and self.other_user_id is None:
            raise ValueError(
                "Per segnalare un messaggio privato serve l'altro utente.",
            )

        if self.category == "private":
            if not self.disclosure_consent:
                raise ValueError(
                    "Per segnalare un messaggio cifrato devi acconsentire a "
                    "renderlo leggibile ai moderatori.",
                )

            if not (self.disclosed_content_key or "").strip():
                raise ValueError(
                    "Chiave di disclosure mancante: il messaggio non "
                    "sarebbe verificabile.",
                )

        if self.category != "private" and self.disclosed_content_key:
            raise ValueError(
                "La chiave di disclosure vale solo per i messaggi privati.",
            )

        return self


class NewsReportModerationRequest(
    BaseModel,
):
    status: NewsReportStatus

    action: NewsReportAction = "none"

    note: str = Field(
        default="",
        max_length=2000,
    )

    @field_validator(
        "note",
    )
    @classmethod
    def normalize_note(
        cls,
        value: str,
    ) -> str:
        return value.strip()


class NewsReportResponse(
    BaseModel,
):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: int
    category: str
    news_id: str
    group_id: int | None
    conversation_id: str | None
    reporter_user_id: int
    reported_user_id: int | None
    reason: str
    description: str | None
    status: str
    moderation_action: str
    moderation_note: str | None
    reviewed_by_user_id: int | None
    reviewed_at: datetime | None
    disclosure_consent_at: datetime | None
    disclosure_opened_at: datetime | None
    created_at: datetime
    updated_at: datetime


class NewsReportListResponse(
    BaseModel,
):
    items: list[NewsReportResponse]
    total: int
    limit: int
    offset: int


class NewsReportDisclosureResponse(
    BaseModel,
):
    report_id: int
    category: str
    news_id: str
    author_id: int | None
    author_name: str
    created_at: datetime | None
    content: str
    verified: bool
    wrap_targets: list[str] = []
